astropy/cosmology/funcs.py

```diff
@@ -90,56 +125,118 @@ def z_at_value(func, fval, zmin=1e-8, zmax=1000, ztol=1e-8, maxfun=500):
 
     Also note that the luminosity distance and distance modulus (two
     other commonly inverted quantities) are monotonic in flat and open
     universes, but not in closed universes.
     """
-    from scipy.optimize import fminbound
-
-    fval_zmin = func(zmin)
-    fval_zmax = func(zmax)
+    from scipy.optimize import minimize_scalar
+
+    opt = {'maxiter': maxfun}
+    # Assume custom methods support the same options as default; otherwise user will see warnings.
+    if str(method).lower() == 'bounded':
+        opt['xatol'] = ztol
+        if bracket is not None:
+            warnings.warn(f"Option 'bracket' is ignored by method {method}.")
+            bracket = None
+    else:
+        opt['xtol'] = ztol
+
+    # fval falling inside the interval of bracketing function values does not guarantee
+    # it has a unique solution, but for Standard Cosmological quantities normally should
+    # (being monotonic or having a single extremum).
+    # In these cases keep solver from returning solutions outside of bracket.
+    fval_zmin, fval_zmax = func((zmin, zmax))
+    nobracket = False
     if np.sign(fval - fval_zmin) != np.sign(fval_zmax - fval):
-        warnings.warn("""\
-fval is not bracketed by func(zmin) and func(zmax). This means either
-there is no solution, or that there is more than one solution between
-zmin and zmax satisfying fval = func(z).""")
+        if bracket is None:
+            nobracket = True
+        else:
+            fval_brac = func(bracket)
+            if np.sign(fval - fval_brac[0]) != np.sign(fval_brac[-1] - fval):
+                nobracket = True
+            else:
+                zmin, zmax = bracket[0], bracket[-1]
+                fval_zmin, fval_zmax = fval_brac[[0, -1]]
+    if nobracket:
+        warnings.warn(f"fval is not bracketed by func(zmin)={fval_zmin} and func(zmax)="
+                      f"{fval_zmax}. This means either there is no solution, or that there is "
+                      "more than one solution between zmin and zmax satisfying fval = func(z).",
+                      AstropyUserWarning)
 
     if isinstance(fval_zmin, Quantity):
         val = fval.to_value(fval_zmin.unit)
-        f = lambda z: abs(func(z).value - val)
     else:
-        f = lambda z: abs(func(z) - fval)
-
-    zbest, resval, ierr, ncall = fminbound(f, zmin, zmax, maxfun=maxfun,
-                                           full_output=1, xtol=ztol)
-
-    if ierr != 0:
-        warnings.warn(f'Maximum number of function calls ({ncall}) reached')
-
-    if np.allclose(zbest, zmax):
-        raise CosmologyError("Best guess z is very close the upper z limit.\n"
+        val = fval
+
+    # 'Brent' and 'Golden' ignore `bounds`, force solution inside zlim
+    def f(z):
+        if z > zmax:
+            return 1.e300 * (1.0 + z - zmax)
+        elif z < zmin:
+            return 1.e300 * (1.0 + zmin - z)
+        elif isinstance(fval_zmin, Quantity):
+            return abs(func(z).value - val)
```

Comment:
While I like that this has less duplicated code, I think having the prior implementation of two functions might still be best.
Adding the z-bounds checks reduces calls to ``func``, which is great, but this adds an extra check every time. Without a side-by-side performance comparison, I image that having one function for non-quantity and one with quantity is faster.

Reply:
The z-bounds checks are inevitable, since some minimisers ignore the `bounds` settings, so they must be forced to reject those solutions, but the speedup is a nice side effect.

I must have misunderstood your earlier comment on the interpretation of the `return abs(func(z).value - val)` branch; but anyway I had checked that the conditional overhead is at the sub-percent level here, e.g. on my old notebook:

- `Planck18.angular_diameter_distance(z)`: 75 µs
- `Planck18.efunc(z)`: 25 µs
- `u.Quantity(x).value`:  4.5 µs
- `return x.value if isinstance(x, u.Quantity) else x`: 720 ns / 200 ns (for quantities / scalars)
- `return x` : 90 ns

so 100-120 ns for evaluating the condition is hardly worth the cost in readability and maintainability imo, since the function has also grown a bit beyond a simple `lambda z:`. And the function evaluations are not that numerous here, typically around 20.

Reply:
SGTM.

Reply:
Though I realised the alternative
`return 1.e300 * (1.0 + z - zmax) if z > zmax else 1.e300 * (1.0 + zmin - z) if z < zmin else abs(func(z).value - val`
just barely does not fit on one line... ;-)

Reply:
That would be nice.

Reply:
Yes, I thought about cutting the values to `1e99` and further abbreviating the parameters, but I must say I did not find the one-liner actually more readable!

Reply:
especially with the nested ternary. Hard to quickly parse.